File: src/board.py
from enum import Enum
from typing import Dict
class Action(Enum):
    LEFT = 0
    RIGHT = 1
    UP = 2
    DOWN = 3

class Board:
    __is_lookup_tables_initialized: bool = False
    __left_moves: list[int] = [0] * (2**16)
    __right_moves: list[int] = [0] * (2**16)
    empty_cells: Dict[int, list[tuple[int, int]]] = {}

    def __init__(self, state: int = None):
        if state is not None:
            Board.verify_state(state=state)
            self.__state = state
        else:
            self.__state = 0
        if not Board.__is_lookup_tables_initialized:
            Board.__init_lookup_tables()
            Board.__is_lookup_tables_initialized = True

    @staticmethod
    def is_lookup_tables_initialized() -> bool:
        return Board.__is_lookup_tables_initialized
    
    @staticmethod
    def _move_left(row: list[int]) -> list[int]:
        # Validate input
        if len(row) != 4:
            raise ValueError("Invalid row length")
        if not all(0 <= x <= 15 for x in row):
            raise ValueError(f"Invalid row values: row = {row}")
        if not isinstance(row, list) or not all(isinstance(x, int) for x in row):
            raise TypeError("Invalid row type")
        non_zero = [x for x in row if x != 0]
        result = [0, 0, 0, 0]
        skip = False
        i = j = 0
        while i < len(non_zero):
            if skip:
                skip = False
                i += 1
                continue
            if i + 1 < len(non_zero) and non_zero[i] == non_zero[i + 1]:
                # Check for 15+15 edge case
                if non_zero[i] == 15:
                    result[j] = non_zero[i]
                    j += 1
                    result[j] = non_zero[i+1]
                    skip = True
                else:
                    result[j] = non_zero[i] + 1
                    skip = True
            else:
                result[j] = non_zero[i]
            i += 1
            j += 1
        return result

    @staticmethod
    def _move_right(row: list[int]) -> list[int]:
        reversed_row = list(reversed(row))
        moved = Board._move_left(reversed_row)
        return list(reversed(moved))

    @staticmethod
    def __init_lookup_tables():
        # Guard against multiple initializations
        if Board.__is_lookup_tables_initialized:
            return
            
        for i in range(2**16):
            row = [
                (i >> 12) & 0xF,
                (i >> 8) & 0xF,
                (i >> 4) & 0xF,
                i & 0xF
            ]
            new_row_left = Board._move_left(row)
            new_value_left = (new_row_left[0] << 12) | (new_row_left[1] << 8) | (new_row_left[2] << 4) | new_row_left[3]
            Board.__left_moves[i] = new_value_left

            new_row_right = Board._move_right(row)
            new_value_right = (new_row_right[0] << 12) | (new_row_right[1] << 8) | (new_row_right[2] << 4) | new_row_right[3]
            Board.__right_moves[i] = new_value_right

    @staticmethod
    def _row_left(row: int) -> int:
        # Verify row.
        if not isinstance(row, int):
            raise TypeError(f"Invalid row type: {row}")
        if row < 0:
            raise IndexError(f"Invalid row: {row}")
        return Board.__left_moves[row]

    @staticmethod
    def _row_right(row: int) -> int:
        if row < 0:
            raise IndexError(f"Invalid row: {row}")
        if row >= (2**16):
            raise IndexError(f"Invalid row: {row}")
        if not isinstance(row, int):
            raise TypeError(f"Invalid row type: {row}")
        if not Board.__is_lookup_tables_initialized:
            Board.__init_lookup_tables()
        return Board.__right_moves[row]

    @staticmethod
    def verify_state(state: int) -> bool:
        if not isinstance(state, int):
            raise TypeError(f"Invalid state type: {state}")
        if state < 0:
            raise ValueError(f"Invalid state: {state}")
        if state >= (2**64):
            raise OverflowError(f"Invalid state: {state}")
        return True

    # @staticmethod
    # def simulate_move(state: int, action: Action) -> int:
    #     # Verify action.
    #     if action not in Action:
    #         raise ValueError(f"Unsupported action: {action}")
    #     new_state = 0
    #     if action in (Action.LEFT, Action.RIGHT):
    #         for row_index in range(4):
    #             row = (state >> (16 * row_index)) & 0xFFFF
    #             if action == Action.LEFT:
    #                 new_row = Board._row_left(row)
    #             else:
    #                 new_row = Board._row_right(row)
    #             new_state |= new_row << (16 * row_index)
    #     elif action in (Action.UP, Action.DOWN):
    #         rows = [(state >> (16 * i)) & 0xFFFF for i in range(4)]
    #         new_rows = [0, 0, 0, 0]
    #         for col in range(4):
    #             col_value = 0
    #             for row_index in range(4):
    #                 tile = (rows[row_index] >> (4 * (3 - col))) & 0xF
    #                 col_value |= tile << (4 * (3 - row_index))
    #             if action == Action.UP:
    #                 new_col_value = Board.__left_moves[col_value]
    #             else:
    #                 new_col_value = Board.__right_moves[col_value]
    #             for row_index in range(4):
    #                 new_tile = (new_col_value >> (4 * (3 - row_index))) & 0xF
    #                 new_rows[row_index] |= new_tile << (4 * (3 - col))
    #         for row_index in range(4):
    #             new_state |= new_rows[row_index] << (16 * row_index)
    #     else:
    #         raise ValueError(f"Unsupported action: {action}")
    #     return new_state

    @staticmethod
    def simulate_moves(state: int) -> list[int]:
        # Verify input
        Board.verify_state(state)

        # Horizontal moves: LEFT and RIGHT.
        new_state_left = new_state_right = 0
        for row_index in range(4):
            row = (state >> (16 * row_index)) & 0xFFFF
            new_row_left = Board.__left_moves[row]
            new_row_right = Board.__right_moves[row]
            new_state_left |= new_row_left << (16 * row_index)
            new_state_right |= new_row_right << (16 * row_index)

        # Vertical moves: reuse the row extraction.
        rows = [(state >> (16 * i)) & 0xFFFF for i in range(4)]
        new_rows_up = [0, 0, 0, 0]
        new_rows_down = [0, 0, 0, 0]
        for col in range(4):
            col_value = 0
            for row_index in range(4):
                tile = (rows[row_index] >> (4 * (3 - col))) & 0xF
                col_value |= tile << (4 * (3 - row_index))
            new_col_value_up = Board.__right_moves[col_value]
            new_col_value_down = Board.__left_moves[col_value]
            for row_index in range(4):
                new_tile_up = (new_col_value_up >> (4 * (3 - row_index))) & 0xF
                new_tile_down = (new_col_value_down >> (4 * (3 - row_index))) & 0xF
                new_rows_up[row_index] |= new_tile_up << (4 * (3 - col))
                new_rows_down[row_index] |= new_tile_down << (4 * (3 - col))
        new_state_up = new_state_down = 0
        for row_index in range(4):
            new_state_up |= new_rows_up[row_index] << (16 * row_index)
            new_state_down |= new_rows_down[row_index] << (16 * row_index)

        return [new_state_left, new_state_right, new_state_up, new_state_down]

    @staticmethod
    def get_valid_move_actions(state: int) -> list[tuple[Action, int]]:
        valid_actions = []
        next_states = Board.simulate_moves(state)
        for action_value, next_state in enumerate(next_states):
            if next_state != state:
                valid_actions.append((Action(action_value), next_state))
        return valid_actions

    @staticmethod
    def get_empty_tiles(state: int) -> list[tuple[int, int]]:
        # Verify input
        Board.verify_state(state)

        if state in Board.empty_cells:
            return Board.empty_cells[state]
        else:
            empty_tiles = []
            for idx, value in enumerate(Board.get_unpacked_state(state)):
                if value == 0:
                    empty_tiles.append((idx // 4, idx % 4))
            Board.empty_cells[state] = empty_tiles
        return empty_tiles

    @staticmethod
    def set_tile(state: int, row: int, col: int, value: int):
        Board.verify_state(state)
        if value < 0 or value > 15:
            raise ValueError("Value must be between 0 and 15.")
        if row < 0 or row > 3 or col < 0 or col > 3:
            raise ValueError("Row and column must be between 0 and 3.")
        # Convert row/col to bit position from left to right
        i = (3 - row) * 4 + (3 - col)
        if ((state >> (i * 4)) & 0xF) != 0:
            raise ValueError("Tile at given row and column is not empty.")
        state |= value << (i * 4)
        return state

    # set_state and get_state stay instance methods.
    def set_state(self, state: int):
        # Verify state.
        Board.verify_state(state)
        self.__state = state

    def get_state(self, unpack: bool = False) -> int | list[int]:
        # Verify unpack is of type bool
        if not isinstance(unpack, bool):
            raise TypeError("unpack must be of type bool.")
        if unpack:
            return Board.get_unpacked_state(self.__state)
        return self.__state

    @staticmethod
    def get_unpacked_state(state: int) -> list[int]:
        Board.verify_state(state)
        return [(state >> (i * 4)) & 0xF for i in range(15,-1,-1)]

    @staticmethod
    def pretty_print(state: int):
        Board.verify_state(state)
        print("-" * 11)
        for i in range(4):
            row = (state >> ((3-i) * 16)) & 0xFFFF
            print("{:X} {:X} {:X} {:X}".format(
                row >> 12, (row >> 8) & 0xF, (row >> 4) & 0xF, row & 0xF))

File: cpp/src/board.cpp
// board.cpp
#include "board.hpp"
#include <algorithm>
#include <array>
#include <iostream>
#include <iomanip>
#include <random>

bool Board::lookupInitialized = false;
std::array<uint16_t, 1 << 16> Board::leftMoves;
std::array<uint16_t, 1 << 16> Board::rightMoves;
std::array<int, 1 << 16> Board::leftScores;
std::array<int, 1 << 16> Board::rightScores;

std::string actionToString(Action action) {
    switch (action) {
        case Action::LEFT:
            return "LEFT";
        case Action::RIGHT:
            return "RIGHT";
        case Action::UP:
            return "UP";
        case Action::DOWN:
            return "DOWN";
        default:
            return "INVALID";
    }
}

uint16_t Board::moveRight(uint16_t row, int& score) {
    int values[4] = {
        (row >> 12) & 0xF,
        (row >> 8) & 0xF,
        (row >> 4) & 0xF,
        row & 0xF
    };

    // Reset score
    score = 0;

    // Remove zeros and merge
    std::vector<int> nonZero;
    for (int val : values) {
        if (val != 0) nonZero.push_back(val);
    }

    std::vector<int> result(4, 0);
    int j = 0;
    for (size_t i = 0; i < nonZero.size(); ++i) {
        if (i + 1 < nonZero.size() && nonZero[i] == nonZero[i + 1]) {
            if (nonZero[i] == 15) {
                result[j++] = nonZero[i];
                result[j++] = nonZero[i];
                ++i;
            } else {
                // When merging, add the score: 2^(value+1)
                int mergedValue = nonZero[i] + 1;
                result[j++] = mergedValue;
                score += valueToTile(mergedValue);
                ++i;
            }
        } else {
            result[j++] = nonZero[i];
        }
    }

    return static_cast<uint16_t>((result[0] << 12) | (result[1] << 8) | (result[2] << 4) | result[3]);
}

uint16_t Board::moveLeft(uint16_t row, int& score) {
    uint16_t reversed = ((row & 0xF) << 12) | ((row & 0xF0) << 4) |
                       ((row & 0xF00) >> 4) | ((row & 0xF000) >> 12);
    uint16_t moved = moveRight(reversed, score);
    return ((moved & 0xF) << 12) | ((moved & 0xF0) << 4) |
           ((moved & 0xF00) >> 4) | ((moved & 0xF000) >> 12);
}

int Board::countEmptyTiles(uint16_t row) {
    int count = 0;
    for (int i = 0; i < 4; ++i) {
        if (((row >> (i * 4)) & 0xF) == 0) {
            count++;
        }
    }
    return count;
}

void Board::initLookupTables() {
    for (int i = 0; i < (1 << 16); ++i) {
        int leftScore = 0;
        int rightScore = 0;
        leftMoves[i] = static_cast<uint16_t>(moveLeft(static_cast<uint16_t>(i), leftScore));
        rightMoves[i] = static_cast<uint16_t>(moveRight(static_cast<uint16_t>(i), rightScore));
        leftScores[i] = leftScore;
        rightScores[i] = rightScore;
    }
}

std::vector<std::tuple<int, int>> Board::getEmptyTiles(BoardState state) {
    std::vector<std::tuple<int, int>> empty;
    empty.reserve(16);  // Pre-allocate maximum possible size

    // Use 16-bit word processing
    for (int row = 0; row < 4; ++row) {
        uint16_t rowVal = (state >> (row * 16)) & 0xFFFF;

        for (int col = 0; col < 4; ++col) {
            if (((rowVal >> (col * 4)) & 0xF) == 0) {
                empty.emplace_back(row, col);
            }
        }
    }
    return empty;
}

BoardState Board::setTile(BoardState state, int row, int col, int value) {
    int pos = (row * 4 + col) * 4;
    return state | (static_cast<BoardState>(value) << pos);
}

std::vector<std::tuple<BoardState, int>> Board::simulateMovesWithScores(BoardState state) {
    std::vector<std::tuple<BoardState, int>> results(4);

    // Initialize all states to the original state and scores to 0
    for (int i = 0; i < 4; ++i) {
        std::get<0>(results[i]) = 0;
        std::get<1>(results[i]) = 0;
    }

    // Vertical moves
    BoardState transposedState = transpose(state);
    // Horizontal moves
    for (int idx = 0; idx < 4; ++idx) {
        BoardState shift = 16 * idx;
        uint16_t rowVal = (state >> shift) & 0xFFFF;
        std::get<0>(results[0]) |= static_cast<BoardState>(leftMoves[rowVal]) << shift;
        std::get<1>(results[0]) += leftScores[rowVal];

        std::get<0>(results[1]) |= static_cast<BoardState>(rightMoves[rowVal]) << shift;
        std::get<1>(results[1]) += rightScores[rowVal];

        uint16_t colVal = (transposedState >> shift) & 0xFFFF;
        std::get<0>(results[2]) |= static_cast<BoardState>(leftMoves[colVal]) << shift;
        std::get<1>(results[2]) += leftScores[colVal];

        std::get<0>(results[3]) |= static_cast<BoardState>(rightMoves[colVal]) << shift;
        std::get<1>(results[3]) += rightScores[colVal];
    }

    std::get<0>(results[2]) = transpose(std::get<0>(results[2]));
    std::get<0>(results[3]) = transpose(std::get<0>(results[3]));

    return results;
}

std::vector<ChosenActionResult> Board::getValidMoveActionsWithScores(BoardState state) {
    std::vector<ChosenActionResult> valid;
    auto moves = simulateMovesWithScores(state);

    for (int i = 0; i < 4; ++i) {
        if (std::get<0>(moves[i]) != state) {
            valid.emplace_back(static_cast<Action>(i), std::get<0>(moves[i]), std::get<1>(moves[i]));
        }
    }

    return valid;
}

std::vector<std::tuple<Action, BoardState>> Board::getValidMoveActions(BoardState state) {
    auto movesWithScores = getValidMoveActionsWithScores(state);
    std::vector<std::tuple<Action, BoardState>> valid;

    for (const auto& [action, nextState, score] : movesWithScores) {
        valid.emplace_back(action, nextState);
    }

    return valid;
}

/**
 * @brief Transposes the board (converts rows to columns and vice versa)
 *
 * Uses efficient bit manipulation to transpose the 4x4 board by rearranging
 * the 4-bit tiles within the 64-bit state.
 *
 * @param state 64-bit board state
 * @return BoardState Transposed board state
 */
BoardState Board::transpose(BoardState state) {
    BoardState a1 = state & 0xF0F00F0FF0F00F0FULL;
    BoardState a2 = state & 0x0000F0F00000F0F0ULL;
    BoardState a3 = state & 0x0F0F00000F0F0000ULL;
    BoardState a = a1 | (a2 << 12) | (a3 >> 12);
    BoardState b1 = a & 0xFF00FF0000FF00FFULL;
    BoardState b2 = a & 0x00FF00FF00000000ULL;
    BoardState b3 = a & 0x00000000FF00FF00ULL;
    return b1 | (b2 >> 24) | (b3 << 24);
}

// Helper function to unpack state into a 2D board
void Board::unpackState(BoardState state, uint8_t board[4][4]) {
    for (int row = 0; row < 4; row++) {
        for (int col = 0; col < 4; col++) {
            board[row][col] = static_cast<uint8_t>(
                (state >> ((row * 4 + col) * 4)) & 0xF
            );
        }
    }
}

void Board::printBoard(uint8_t board[4][4]) {
    std::cout << "+------+------+------+------+" << std::endl;
    for (int i = 0; i < 4; ++i) {
        for (int j = 0; j < 4; ++j) {
            std::cout << "| " << std::setw(5) << (board[i][j] ? std::to_string(board[i][j]) : " ");
        }
        std::cout << "|" << std::endl;
        std::cout << "+------+------+------+------+" << std::endl;
    }
}

BoardState Board::randomizeState() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> probDist(0.0, 1.0);

    BoardState state = 0;
    // There are 16 tiles on the board
    for (int i = 0; i < 16; ++i) {
        double r = probDist(gen);
        int tile;
        if (r < 0.5) {
            tile = 0;
        } else if (r < 0.8) { // 0.3 probability
            std::uniform_int_distribution<int> dist(1, 10);
            tile = dist(gen);
        } else if (r < 0.95) { // 0.15 probability
            std::uniform_int_distribution<int> dist(11, 12);
            tile = dist(gen);
        } else { // remaining 0.05 probability
            std::uniform_int_distribution<int> dist(13, 15);
            tile = dist(gen);
        }
        // Each tile is stored in 4 bits in the 64-bit state.
        state |= (static_cast<BoardState>(tile) << (i * 4));
    }

    return state;
}
